<a href="https://colab.research.google.com/github/emekeh/HDSC-Winter-23/blob/main/Stage_A_Quiz1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Hamoye Stage A Quiz**

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
# Importing the necessary python libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [ ]:
# Import dataset
df = pd.read_csv("/content/drive/MyDrive/files/stage_A_dataset.csv", encoding="latin-1")
df.head()

Area Code     Area  Item Code         Item  Element Code  \
0          4  Algeria       2501   Population           511   
1          4  Algeria       2501   Population          5301   
2          4  Algeria       2901  Grand Total           664   
3          4  Algeria       2901  Grand Total           674   
4          4  Algeria       2901  Grand Total           684   

                                  Element             Unit     Y2014  \
0           Total Population - Both sexes     1000 persons  38924.00   
1                Domestic supply quantity      1000 tonnes      0.00   
2           Food supply (kcal/capita/day)  kcal/capita/day   3377.00   
3  Protein supply quantity (g/capita/day)     g/capita/day     94.90   
4      Fat supply quantity (g/capita/day)     g/capita/day     80.06   

      Y2015     Y2016     Y2017     Y2018  
0  39728.00  40551.00  41389.00  42228.00  
1      0.00      0.00      0.00      0.00  
2   3379.00   3372.00   3341.00   3322.00  
3     94.35     94.72     92.82     91.83  
4     79.36     77.40     80.19     77.28

In [ ]:
# check for missing values
df.isnull().sum()

Area Code          0
Area               0
Item Code          0
Item               0
Element Code       0
Element            0
Unit               0
Y2014           1589
Y2015           1548
Y2016           1535
Y2017           1506
Y2018           1436
dtype: int64

In [ ]:
# view the shape of the dataset
print(df.shape)
# view summary of the information
print(df.info())

(60943, 12)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 60943 entries, 0 to 60942
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Area Code     60943 non-null  int64  
 1   Area          60943 non-null  object 
 2   Item Code     60943 non-null  int64  
 3   Item          60943 non-null  object 
 4   Element Code  60943 non-null  int64  
 5   Element       60943 non-null  object 
 6   Unit          60943 non-null  object 
 7   Y2014         59354 non-null  float64
 8   Y2015         59395 non-null  float64
 9   Y2016         59408 non-null  float64
 10  Y2017         59437 non-null  float64
 11  Y2018         59507 non-null  float64
dtypes: float64(5), int64(3), object(4)
memory usage: 5.6+ MB
None


In [ ]:
# Percentage of missing data
df.isnull().count()

Area Code       60943
Area            60943
Item Code       60943
Item            60943
Element Code    60943
Element         60943
Unit            60943
Y2014           60943
Y2015           60943
Y2016           60943
Y2017           60943
Y2018           60943
dtype: int64

In [ ]:
# Summary statistics
df.describe()

Area Code     Item Code  Element Code          Y2014          Y2015  \
count  60943.000000  60943.000000  60943.000000   59354.000000   59395.000000   
mean     134.265576   2687.176706   3814.856456     134.196282     135.235966   
std       72.605709    146.055739   2212.007033    1567.663696    1603.403984   
min        4.000000   2501.000000    511.000000   -1796.000000   -3161.000000   
25%       74.000000   2562.000000    684.000000       0.000000       0.000000   
50%      136.000000   2630.000000   5142.000000       0.090000       0.080000   
75%      195.000000   2775.000000   5511.000000       8.340000       8.460000   
max      276.000000   2961.000000   5911.000000  176405.000000  181137.000000   

               Y2016          Y2017          Y2018  
count   59408.000000   59437.000000   59507.000000  
mean      136.555222     140.917765     143.758381  
std      1640.007194    1671.862359    1710.782658  
min     -3225.000000   -1582.000000   -3396.000000  
25%         0.000000       0.000000       0.000000  
50%         0.080000       0.100000       0.070000  
75%         8.430000       9.000000       9.000000  
max    185960.000000  190873.000000  195875.000000

In [ ]:
# Standard deviation
df.std()

<ipython-input-8-ed60dc45d152>:2: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  df.std()


Area Code         72.605709
Item Code        146.055739
Element Code    2212.007033
Y2014           1567.663696
Y2015           1603.403984
Y2016           1640.007194
Y2017           1671.862359
Y2018           1710.782658
dtype: float64

In [ ]:
# Mean
df.mean()

<ipython-input-9-a3576ddd3f19>:2: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  df.mean()


Area Code        134.265576
Item Code       2687.176706
Element Code    3814.856456
Y2014            134.196282
Y2015            135.235966
Y2016            136.555222
Y2017            140.917765
Y2018            143.758381
dtype: float64

In [ ]:
# The sum of each `"Item"` per year
df.groupby(["Item"]).sum()

Area Code  Item Code  Element Code      Y2014      Y2015  \
Item                                                                            
Alcohol, Non-Food        38447     752497       1515090    2403.00    2180.00   
Alcoholic Beverages      75016    1649136       2171381  102410.11   98783.72   
Animal Products          18060     397035         90990   11935.65   11811.73   
Animal fats              85279    1882494       2567303  209460.54  200675.72   
Apples and products      69280    1352989       1920418    9499.23   10559.15   
...                        ...        ...           ...        ...        ...   
Vegetables, Other        77779    1518715       2275409  155038.96  158104.08   
Vegetal Products         18060     391905         90990  107145.19  107064.17   
Wheat and products       87073    1639683       2654934  232670.13  234710.51   
Wine                     66206    1319535       1818328    4497.36    4251.81   
Yams                     50798     970905       1525745  200396.96  203151.78   

                         Y2016      Y2017      Y2018  
Item                                                  
Alcohol, Non-Food      2223.00    2348.00    2293.00  
Alcoholic Beverages   96958.75   95581.06   97847.27  
Animal Products       11661.69   11547.65   11578.61  
Animal fats          183314.22  269617.53  269648.27  
Apples and products    9853.26   10198.90    9640.51  
...                        ...        ...        ...  
Vegetables, Other    156218.90  157752.59  163987.21  
Vegetal Products     106792.72  107655.20  107775.39  
Wheat and products   239377.07  240047.62  242645.19  
Wine                   3872.09    4178.02    4039.32  
Yams                 220626.20  229174.59  221272.09  

[119 rows x 8 columns]

In [ ]:
# Total sum of each `"Element"` per year
df.groupby("Element").sum()

Area Code  Item Code  Element Code  \
Element                                                                      
Domestic supply quantity                   708993   14197445      28068795   
Export Quantity                            599910   11840553      26026133   
Fat supply quantity (g/capita/day)         675050   13535000       3435732   
Feed                                       176272    3538507       7282199   
Food                                       663295   13285035      25406622   
Food supply (kcal/capita/day)              674057   13511060       3329296   
Food supply quantity (kg/capita/yr)        658446   13185401       3163725   
Import Quantity                            688174   13795966      28834929   
Losses                                     274353    5424803      10292107   
Other uses (non-food)                      235554    4729749       8926728   
Processing                                 271940    5350416      10313310   
Production                                 526751   10450053      21388191   
Protein supply quantity (g/capita/day)     675050   13535000       3385502   
Residuals                                  623271   12421089      24066350   
Seed                                       103537    2035933       4211574   
Stock Variation                            571566   11329527      21464704   
Total Population - Both sexes                6020     112545         22995   
Tourist consumption                         50308    1486528       2869905   

                                             Y2014       Y2015       Y2016  \
Element                                                                      
Domestic supply quantity                1996716.35  2021493.55  2044842.70   
Export Quantity                          150020.64   157614.47   151920.46   
Fat supply quantity (g/capita/day)        10225.56    10235.74    10102.77   
Feed                                     216927.89   225050.22   228958.65   
Food                                    1212332.49  1232361.10  1247022.17   
Food supply (kcal/capita/day)            454257.00   453383.00   451810.00   
Food supply quantity (kg/capita/yr)       49650.63    49345.13    48985.28   
Import Quantity                          274144.48   267018.46   286582.78   
Losses                                   153223.00   155439.00   157787.00   
Other uses (non-food)                     78718.13    66254.41    69563.68   
Processing                               282923.00   287929.00   280631.00   
Production                              1931287.75  1947019.39  1943537.15   
Protein supply quantity (g/capita/day)    11836.46    11833.95    11779.69   
Residuals                                 30149.00    30045.00    37224.00   
Seed                                      21922.92    23976.82    23389.20   
Stock Variation                           58749.83    34910.99    33140.12   
Total Population - Both sexes           1031585.00  1058081.00  1085107.00   
Tourist consumption                         416.00      349.00       89.00   

                                             Y2017       Y2018  
Element                                                         
Domestic supply quantity                2088198.10  2161192.10  
Export Quantity                          182338.80   181594.80  
Fat supply quantity (g/capita/day)        10253.84    10258.69  
Feed                                     223705.68   233489.68  
Food                                    1258888.28  1303841.28  
Food supply (kcal/capita/day)            454681.00   455261.00  
Food supply quantity (kg/capita/yr)       48690.04    49056.85  
Import Quantity                          294559.09   287997.09  
Losses                                   160614.00   163902.00  
Other uses (non-food)                     91645.97    91300.97  
Processing                               292836.00   308429.00  
Production                              2030056.89  2075072.89  
Protein 

In [ ]:
# Total sum according to `"Area"` in 2017
df.groupby("Area")[["Y2017"]].sum().sort_values(by="Y2017", ascending=False)

Y2017
Area                                   
Nigeria                      1483268.23
Egypt                         866379.92
South Africa                  517590.54
Ethiopia                      448683.76
Morocco                       388495.36
Ghana                         337599.06
Algeria                       325644.27
United Republic of Tanzania   322616.85
Kenya                         264660.66
Sudan                         239931.92
Cameroon                      232030.43
Angola                        229159.57
Côte d'Ivoire                 224599.01
Uganda                        213950.38
Malawi                        181098.71
Mozambique                    161407.98
Mauritania                    156665.46
Mali                          149928.33
Madagascar                    131197.73
Niger                         126707.58
Benin                         124771.22
Tunisia                       124167.20
Zambia                        103223.77
Burkina Faso                  101855.07
Guinea                         98138.87
Senegal                        95681.15
Zimbabwe                       75919.34
Rwanda                         73663.69
Chad                           71594.68
Sierra Leone                   55311.33
Eswatini                       54343.33
Mauritius                      51114.83
Togo                           49841.88
Congo                          41181.68
Central African Republic       29937.00
Namibia                        29874.89
Liberia                        29342.20
Gabon                          27979.64
Gambia                         23154.18
Djibouti                       22729.91
Botswana                       22101.30
Lesotho                        21267.96
Guinea-Bissau                  19102.77
Cabo Verde                     14650.74
Sao Tome and Principe          12662.63
Seychelles                       442.34
Comoros                           59.84
Sudan (former)                     0.00
Ethiopia PDR                       0.00

### **Question 1:** Which of these Areas had the highest sum in 2017?

In [ ]:
# Area with the highest sum in 2017
df.groupby("Area")[["Y2017"]].sum().sort_values(by="Y2017", ascending=False).iloc[0]

Y2017    1483268.23
Name: Nigeria, dtype: float64

### **Question 2:** What is the **mean** and **standard deviation** across the whole dataset for the year 2017 to 2 decimal places?

In [ ]:
# Mean and standard deviation across the whole dataset in 2017.
mean = df["Y2017"].mean()

std = df["Y2017"].std()

print(f"Mean for Year 2017: {round(mean, 3)}")

print(f"Standard Deviation for Year 2017: {round(std, 3)}")

Mean for Year 2017: 140.918
Standard Deviation for Year 2017: 1671.862


### **Question 3:** What is the total number of the sum of `"Processing"` in 2017?

In [ ]:
# total number of the sum of Processing in 2017
proc_2017 = df.groupby("Element").sum().loc["Processing"]["Y2017"]
print(f'Sum of Processing in 2017 is {round(proc_2017, 2)}')

Sum of Processing in 2017 is 292836.0


### **Question 6:** Which year had the least correlation with `"Element Code"`?

In [ ]:
# Correlation 
df.corr()["Element Code"].sort_values(ascending=False)

Element Code    1.000000
Y2014           0.024457
Y2018           0.024279
Y2017           0.024254
Y2015           0.023889
Y2016           0.023444
Area Code      -0.000209
Item Code      -0.024683
Name: Element Code, dtype: float64

In [ ]:
# The highest correlation with Element Code
df.corr()["Element Code"]["Y2014":].sort_values(ascending=False)

Y2014    0.024457
Y2018    0.024279
Y2017    0.024254
Y2015    0.023889
Y2016    0.023444
Name: Element Code, dtype: float64

ANS: The Year with the least correlation with `"Element Code"`: **2016**

### **Question 8:**  What year has the highest sum of `"Stock Variation"`?

In [ ]:
df.groupby("Element").sum().loc["Stock Variation"]

Area Code         571566.00
Item Code       11329527.00
Element Code    21464704.00
Y2014              58749.83
Y2015              34910.99
Y2016              33140.12
Y2017              54316.91
Y2018              20577.91
Name: Stock Variation, dtype: float64

ANS: The Year that has the highest sum of `"Stock Variation"`: **2014**

### **Question 10:** What is the total `"Protein supply quantity"` in `"Madagascar"` in 2015?

In [ ]:
# Total Protein Supply Quantity in 2015
df.set_index("Area").loc["Madagascar"].groupby("Element")[["Y2015"]].sum()

Y2015
Element                                         
Domestic supply quantity                31214.98
Export Quantity                           494.75
Fat supply quantity (g/capita/day)         91.85
Feed                                     2070.22
Food                                    21120.65
Food supply (kcal/capita/day)            7685.00
Food supply quantity (kg/capita/yr)       871.59
Import Quantity                          1721.80
Losses                                   2176.00
Other uses (non-food)                    3957.12
Processing                               1803.00
Production                              29482.89
Protein supply quantity (g/capita/day)    173.05
Residuals                                -616.00
Seed                                      699.00
Stock Variation                          -505.00
Total Population - Both sexes           24234.00

ANS: The total `"Protein supply quantity"` in Madagascar in 2015 is **173.05**.

### **Question 11:** Which of these Areas had the 7th lowest sum in 2017?

In [ ]:
# Area with the 7th lowest sum in 2017
df.groupby("Area")[["Y2017"]].sum().sort_values(by="Y2017", ascending=False).iloc[-7]

Y2017    19102.77
Name: Guinea-Bissau, dtype: float64

### **Question 12:** What is the total sum of `"Wine"` produced in 2015 and 2018 respectively?

In [ ]:
# Total sum of Wine produced in 2015 and 2018 respectively
df.groupby("Item")[["Y2015", "Y2018"]].sum()

Y2015      Y2018
Item                                     
Alcohol, Non-Food      2180.00    2293.00
Alcoholic Beverages   98783.72   97847.27
Animal Products       11811.73   11578.61
Animal fats          200675.72  269648.27
Apples and products   10559.15    9640.51
...                        ...        ...
Vegetables, Other    158104.08  163987.21
Vegetal Products     107064.17  107775.39
Wheat and products   234710.51  242645.19
Wine                   4251.81    4039.32
Yams                 203151.78  221272.09

[119 rows x 2 columns]

In [ ]:
# Total Sum of Wine produced in 2015 and 2018
wine_2015 = df.groupby("Item")[["Y2015", "Y2018"]].sum().loc["Wine"][0]

wine_2018 = df.groupby("Item")[["Y2015", "Y2018"]].sum().loc["Wine"][1]

print(f"Total sum of Wine produced in 2015: {round(wine_2015, 2)}")

print(f"Total sum of Wine produced in 2018: {round(wine_2018, 2)}")

Total sum of Wine produced in 2015: 4251.81
Total sum of Wine produced in 2018: 4039.32


### **Question 15:** What is the total number and percentage of missing data in 2014 to 3 decimal places?

In [ ]:
# Total number and Percentage of missing data in 2014
null_2014 = df["Y2014"].isnull().sum()
total_2014 = len(df["Y2014"])
null_percent = (null_2014/total_2014) * 100

print(f"Total missing data: {null_2014} \nPercent of missing data: {round(null_percent, 3)}")

Total missing data: 1589 
Percent of missing data: 2.607


### **Question 17:** What is the total number of unique countries in the dataset?

In [ ]:
df["Area"].nunique()

49